coroutines/src/commonMain/kotlin/SynchronizedMap.kt

```diff
@@ -0,0 +1,115 @@
+package com.juul.tuulbox.coroutines
+
+import kotlinx.atomicfu.locks.ReentrantLock
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
+
+/** A [MutableMap] where all reads and writes are protected by a [ReentrantLock]. */
+public class SynchronizedMap<K, V>(
+    private val inner: MutableMap<K, V>,
+) : MutableMap<K, V> {
+
+    private inner class Entry(
+        val entry: MutableMap.MutableEntry<K, V>,
+    ) : MutableMap.MutableEntry<K, V> {
+        override val key: K get() = entry.key
+        override val value: V get() = entry.value
+
+        override fun setValue(newValue: V): V =
+            lock.withLock { entry.setValue(newValue) }
+    }
+
+    @PublishedApi
+    internal val lock: ReentrantLock = reentrantLock()
+
+    override val size: Int
+        get() = lock.withLock { inner.size }
+
+    override fun containsKey(key: K): Boolean =
+        lock.withLock { inner.containsKey(key) }
+
+    override fun containsValue(value: V): Boolean =
+        lock.withLock { inner.containsValue(value) }
+
+    override fun get(key: K): V? =
+        lock.withLock { inner[key] }
+
+    override fun isEmpty(): Boolean =
+        lock.withLock { inner.isEmpty() }
+
+    override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
+        get() = lock.withLock {
+            inner.entries.asSequence()
+                .map { entry -> Entry(entry) as MutableMap.MutableEntry<K, V> }
+                .toMutableSet()
+                .readOnly()
+        }
+
+    override val keys: MutableSet<K>
+        get() = lock.withLock { inner.keys.readOnly() }
+
+    override val values: MutableCollection<V>
+        get() = lock.withLock { inner.values.readOnly() }
+
+    override fun clear() {
+        lock.withLock { inner.clear() }
+    }
+
+    override fun put(key: K, value: V): V? =
+        lock.withLock { inner.put(key, value) }
+
+    override fun putAll(from: Map<out K, V>) {
+        lock.withLock { inner.putAll(from) }
+    }
+
+    override fun remove(key: K): V? =
+        lock.withLock { inner.remove(key) }
+
+    /**
+     * Performs [action] with this map's [ReentrantLock] locked. Useful for when correctness dictates that unlocks should not
+     * occur between multiple calls, such as when calling [MutableMap.getOrPut].
+     */
+    public inline fun <T> synchronized(action: SynchronizedMap<K, V>.() -> T): T = lock.withLock { action(this) }
```

Comment:
Does this mean that library consumers will still need to implement `getOrPut` themselves?

Reply:
`getOrPut` is defined as an extension function, `MutableMap<K, V>.getOrPut`, so it cannot be overwrote. Consumers of `SynchronizedMap` will just have to do `myMap.synchronized { getOrPut` instead of `myMap.getOrPut` if they want to guarantee that the read and write happen atomically.

Reply:
That's reasonable.

Can you add an example code snippet to `synchronized`'s KDoc, please?